server/src/main/java/org/opensearch/index/store/remote/utils/cache/SegmentedCache.java

```diff
@@ -0,0 +1,331 @@
+            usage += c.usage();
+            activeUsage += c.activeUsage();
+        }
+        return new CacheUsage(usage, activeUsage);
+    }
+
+    @Override
+    public CacheStats stats() {
+        long hitCount = 0L;
+        long missCount = 0L;
+        long removeCount = 0L;
+        long removeWeight = 0L;
+        long replaceCount = 0L;
+        long evictionCount = 0L;
+        long evictionWeight = 0L;
+
+        for (RefCountedCache<K, V> cache : table) {
+            CacheStats c = cache.stats();
+            hitCount += c.hitCount();
+            missCount += c.missCount();
+            removeCount += c.removeCount();
+            removeWeight += c.removeWeight();
+            replaceCount += c.replaceCount();
+            evictionCount += c.evictionCount();
+            evictionWeight += c.evictionWeight();
+        }
+        return new CacheStats(hitCount, missCount, removeCount, removeWeight, replaceCount, evictionCount, evictionWeight);
+    }
+
+    enum SingletonWeigher implements Weigher<Object> {
+        INSTANCE;
+
+        @Override
+        public long weightOf(Object value) {
+            return 1;
+        }
+    }
+
+    /**
+     * A listener that ignores all notifications.
+     */
+    enum DiscardingListener implements RemovalListener<Object, Object> {
+        INSTANCE;
+
+        @Override
+        public void onRemoval(RemovalNotification<Object, Object> notification) {
+
+        }
+    }
+
+    /**
+     * @return the capacity per internal segment
+     */
+    public long getPerSegmentCapacity() {
+        return perSegmentCapacity;
+    }
+
+    /**
+     * @return the weigher used for IndexInput
+     */
+    public Weigher<V> getWeigher() {
+        return weigher;
+    }
+
+    /**
+     * A builder that creates {@link SegmentedCache} instances. It
+     * provides a flexible approach for constructing customized instances with
+     * a named parameter syntax.
+     */
+    public static final class Builder<K, V> {
+
+        static final int DEFAULT_CONCURRENCY_LEVEL = Runtime.getRuntime().availableProcessors();
+
+        RemovalListener<K, V> listener;
+        Weigher<V> weigher;
+
+        int concurrencyLevel;
+
+        long capacity;
+
+        @SuppressWarnings("unchecked")
+        Builder() {
+            capacity = -1;
+            weigher = (Weigher<V>) SingletonWeigher.INSTANCE;
+            concurrencyLevel = DEFAULT_CONCURRENCY_LEVEL;
+            listener = (RemovalListener<K, V>) DiscardingListener.INSTANCE;
+        }
+
+        /**
+         * Specifies the maximum weighted capacity to coerce the map to and may
+         * exceed it temporarily.
+         *
+         * @param capacity the weighted threshold to bound the map by
+         * @throws IllegalArgumentException if the maximumWeightedCapacity is
+         *                                  negative
+         */
+        public Builder<K, V> capacity(long capacity) {
+            checkArgument(capacity >= 0);
```

Comment:
if (capacity < 0) {
    throw new IllegalArgumentException("The capacity has to be greater or equal to 0");
}